services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java

```diff
@@ -441,7 +442,8 @@ private Response doBrowserLogout(AuthenticationSessionModel logoutSession) {
         // authenticate identity cookie, but ignore an access token timeout as we're logging out anyways.
         AuthenticationManager.AuthResult authResult = AuthenticationManager.authenticateIdentityCookie(session, realm, false);
         if (authResult != null) {
-            userSession = userSession != null ? userSession : authResult.getSession();
+            userSession = userSession != null ? userSession : 
+                    lockObjectsForModification(session, () -> session.sessions().getUserSession(realm, authResult.getSession().getId()));
```

Comment:
When `userSession` is null it's used the one from `authResult`, but that one isn't locked. Therefore new call with locking is being done.

Reply:
Locking of rows in the database works only when the entity is read from the database in the session for the first time. 
When it is locked later, the session might already have a stale entry in its persistence context. 

Therefore, the `lockObjectsForModification()` should wrap the `AuthenticationManager.authenticateIdentityCookie()`. This way the `userSession` embodied in there will be locked when read from the database the first time. 
No further locking is necessary, so the other `lockObjectsForModification()` can then be removed.

Reply:
Thank you, incorporated.